### ISSUE TITLE
IDE0370 false positive, when initializing auto-property with `null!`.

### ISSUE THREAD

**Issue by PetSerAl** [dotnet/roslyn#81843]
**Version Used**:
Locally: SDK 10.0.101
lab.razor.fyi: 5.0.0 and 5.3.0-2.25619.2

**Steps to Reproduce**:
```csharp
class C
{
    public C() { }
    public string S { get; set; } = null!; //IDE0370: Suppression is unnecessary 
}
```
https://lab.razor.fyi/#42pi5OIKKMpPL0rM1UsuFipJzkksLlZw5uWq5uVSUFBQKChNyslMVnDW0FSoVqhFESsuKcrMS1cIVqhWSE8tsVYoBhG1CrYKeaU5OYrWCvr6ni6uBsbmBlYKwaUFBUWpxcWZ-XkKmcUKpXl5qcmpxcWJRZUKvFy1XmypRUX5RcVRnBxvfpw5-SFNgCWBsYIRAA

https://github.com/user-attachments/assets/ca938c88-90ad-4b95-b290-76fffcffa7aa

IDE0370 on `!` sign. 

**Diagnostic Id**:
IDE0370: Suppression is unnecessary

**Expected Behavior**:
No diagnostic should be produced, as `null` assigned to not nullable property.

**Actual Behavior**:
False positive IDE0370. When proposed fix applied, it result in "CS8625: Cannot convert null literal to non-nullable reference type." warning.
No diagnostic, if constructor removed/commented.

**Comment by CyrusNajmabadi** [comment]
Something weird is happening here.  We call:

```
            var updatedDiagnostics = updatedSemanticModel.GetDiagnostics(groupSpan, cancellationToken);
```

With the span of the property.  And we get nothing back.  But if i call GetDiagnostics on the full semantic model, then it returns that diagnostic within that span.

**Comment by CyrusNajmabadi** [comment]
Only repros if you have: 

```
    public C() { }
```

**Comment by CyrusNajmabadi** [comment]
Seems to be an issue with GetDiagnostics when passed a span.  Created repro and have pulled in compiler dev to see if we can figure out waht's up.

**Comment by CyrusNajmabadi** [comment]
ok.  Have verified that hte compiler doesn' tproperly return warnings whne asking for diagnostics for the span for, say, a full property declaration.

It also does not return anything here even if you first ask for diagnostics for a constructor in the type.  if you do ask for diagnostics for the type-declaratio span itself, these will show up.  But that's not super desirable given that a type declaration may be super large itself.

ideally, from talking to @RikkiGibson , the compiler should see that the span is within a property decl and do the appropriate binding of constructors to make nullable warnings show up.

--

If, however, we can't get the GetDiagnostics side fixed, a poor-man's solution here is to update the analyzer to instead bind the entire type declaration, then filter the diagnostics to those that intersect the field/property.

**Comment by RikkiGibson** [comment]
This is a compiler bug. I can hopefully get to this ~~next week~~. It looks like Cyrus has a workaround, and, the gap may be more expansive than we realized at first blush. So, it may be somewhat longer before we are able to get a real fix in on the compiler side.

The path in `MethodCompiler.CompileNamedType`, which simply binds the field initializers, doesn't actually produce the nullable diagnostics for the field initializers.

Instead, nullable analysis of field initializers is delayed until we encounter a constructor which needs to use the "after initializers" state as the constructor's initial state:

https://github.com/dotnet/roslyn/blob/a291525e7ddc85f984d7fcaae259b46967df9f0b/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs#L2115

Whether the nullable warnings are reported or not for a given span, ends up depending on whether the constructor's span intersects with the filterSpan being passed in. I think when the constructor is implicitly declared, then the whole type span is used, which intersects with the filterSpan in this scenario. So we visit that constructor and therefore nullable analyze the field initializers. When the constructor is explicitly declared, the spans don't intersect, so we don't do that nullable analysis.

I believe the fix here is to simply do the nullable analysis of the field initializers, in the same path which binds the field initializers, and store the "after initializers" nullable state on the `Binder.ProcessedFieldInitializers` to use in constructors when we need it.

**Comment by CyrusNajmabadi** [comment]
Reactivating so we track a real fix in the compiler.  @RikkiGibson if you end up fixing this, consider rolling back https://github.com/dotnet/roslyn/pull/81899 as it should not be necessary.

**Comment by bkoelman** [comment]
Looks like this broke again in SDK 10.0.200, but only in command-line builds. It only shows up in VS when the Error List is set to **Build Only**.

```
...\UnusedSelectorsRewriter.cs(33,42): error IDE0370: Suppression is unnecessary (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0370)
```

`SelectNode` is a class, as demonstrated in the screenshot below. The project builds against .NET 8/9/10, so NRT is always available.

<img width="763" height="405" alt="Image" src="https://github.com/user-attachments/assets/11ed02fc-aaf6-441d-b4d9-28157e9c7f15" />

And there's no way to invoke the fixer in the editor window.

<img width="595" height="412" alt="Image" src="https://github.com/user-attachments/assets/e4b19756-4c12-4c75-8c9f-8771282d3101" />

**Comment by CyrusNajmabadi** [comment]
> Looks like this broke again in SDK 10.0.200,

As mentioned above, we only put in a minor heuristic to help. A real fix has to come from the compiler. The bug is there.

**Comment by lcsondes** [comment]
I'd like to mention that this isn't limited to members. With v10.0.200, I'm seeing false-positive IDE0370s for expressions within method bodies, such as `Path.Combine(Environment.GetEnvironmentVariable("ProgramFiles(x86)")!, "...")`, or `return null!;` in a method that returns `object`.

I haven't managed to reproduce these standalone. The Path.Combine example occurs in a Microsoft.Build.Utilities.Task-derived class that does not have a user-defined constructor. return null!'s class has one, single-parameter constructor.

**Comment by CyrusNajmabadi** [comment]
Given that the compiler isn't properly idempotent with nullable analysis, this doesn't surprise me.

**Comment by bkoelman** [comment]
> > Looks like this broke again in SDK 10.0.200,
> 
> As mentioned above, we only put in a minor heuristic to help. A real fix has to come from the compiler. The bug is there.

My point is that *something* changed in SDK 10.0.200: Code that compiled without errrors reproducibly fails when built using this SDK. Is that expected?

**Comment by CyrusNajmabadi** [comment]
> Is that expected?

As the compiler is buggy here, I would not be surprised to see brokenness.

**Comment by RikkiGibson** [comment]
Writing down my plan for fixing this here as I haven't had time to really dig into it yet and I don't want to forget.

1) Nullable-analyze field initializers when they are bound. Store the "after initializers state" on the initializers struct (we have one for instance field initializers and one for static field initializers).
2) When getting an initial nullable state for constructors, do not cook one up each time based on the constructor initializer (checking SetsRequiredMembers and all sorts of stuff). Instead just get/create the one initial state stored on the initializers struct.
3) A constructor initializer is included in the bound constructor body (IIRC). When we visit it in nullable analysis, just look at the constructor and handle things like `: this()` or `[SetsRequiredMembers]` as simply being postconditions on the constructor we are invoking, setting fields to non-null as appropriate.

This should let us "smooth out" a lot of things, reduce repeated work, and improve correctness in this scenario.

**Comment by cremor** [comment]
> Reactivating so we track a real fix in the compiler.

Could someone please update the title (and maybe also the description) of this issue to describe the compiler bug?

**Comment by CyrusNajmabadi** [comment]
> Could someone please update the title (and maybe also the description) of this issue to describe the compiler bug?

The compiler bug is that it doesn't report a warning diagnostic when you do go `string X { get; set; } = null;` because of that, we think the `!` is safe to drop off of `string X { get; set; } = null!;` @cremor

**Comment by cremor** [comment]
I understood that after reading the discussion. What I meant is:
- The problem "IDE0370 false positive, when initializing auto-property with `null!`" was fixed by #81899, correct? So it doesn't make any sense that an issue with that title is still open.
- It would be great if the title (and description) of this issue would be updated to reflect the full scope of the compiler problem. E.g. there are many cases where the analyzer misses an unnecessary suppression.

**Comment by jjonescz** [comment]
I wouldn't edit this issue, it's probably good idea to preserve it. It would be acceptable to open a separate issue though, focused on the compiler issue, and then close this one.

**Comment by CyrusNajmabadi** [comment]
> correct? 

No.  It was mitigated.  For one thing, it won't work properly in some cases with partial types.  A true full fix is needed, and that needs to come from teh compiler.

> It would be great if the title (and description) of this issue would be updated to reflect the full scope of the compiler problem. 

I'd either open a new issue.  Or keep this as is.  The discussion within is appropriate imo.  I would personally not close this.  Because even if one particular case is mitigated, the general issue persists, is known, and is something the compiler team is using to track fixing when they have time.

**Comment by cremor** [comment]
> I wouldn't edit this issue, it's probably good idea to preserve it. It would be acceptable to open a separate issue though, focused on the compiler issue, and then close this one.

> I'd either open a new issue. Or keep this as is. The discussion within is appropriate imo. I would personally not close this.

Ok, this is fine too.
Could one of you please create this issue? I think you can better describe it than me.

**Comment by CyrusNajmabadi** [comment]
I'd prefer to just keep as is.  The description above, and rikki's analysis are already good.
